### PR TITLE
@babel/parser(ts): Add parsing of type import

### PR DESCRIPTION
**Pull request by armano2**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | #7749
| Patch: Bug Fix?          | 👎 
| Major: Breaking Change?  | 👎 
| Minor: New Feature?      | 👍 
| Tests Added + Pass?      | 👍
| Documentation PR Link    | 👎 
| Any Dependency Changes?  | 👎 
| License                  | MIT

This PR is continuation of work done in #8798 by @Kovensky few months ago.
This PR adds new node `TSImportType`.

Comparing to how TS parses this there is one difference
```ts
typeof import("./a")
```
in typescript is represented by `ImportType[typeof=true]`
this implementation will parse it into: `TSTypeQuery[exprName='TSImportType']`

i can easily change that if you think it will be better.

### CODE REVIEW

**Comment by armano2** [comment]
@existentialism you guys have to fix developing on windows, this is killing me xd

finally its finished xd

File: packages/babel-parser/src/plugins/typescript.js [hunk 1]
```diff
@@ -220,6 +220,26 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
+        throw this.unexpected();
+      }
+      node.argument = this.parseLiteral(this.state.value, "StringLiteral");
+      this.expect(tt.parenR);
+
+      if (this.match(tt.dot)) {
+        this.eat(tt.dot);
+        node.qualifier = this.tsParseEntityName(allowReservedWords);
+      }
+      if (!this.hasPrecedingLineBreak() && this.isRelational("<")) {
```
Comment (DanielRosenwasser) [inline packages/babel-parser/src/plugins/typescript.js]: TypeScript doesn't actually check for a preceding line break; do you do similar things throughout Babylon?

Comment (armano2) [reply packages/babel-parser/src/plugins/typescript.js]: that's good point i will change that

File: packages/babel-parser/src/plugins/typescript.js [hunk 2]
```diff
@@ -220,6 +220,26 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
+      this.expect(tt._import);
+      this.expect(tt.parenL);
+      if (!this.match(tt.string)) {
+        throw this.unexpected();
+      }
+      node.argument = this.parseLiteral(this.state.value, "StringLiteral");
+      this.expect(tt.parenR);
+
+      if (this.match(tt.dot)) {
+        this.eat(tt.dot);
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/plugins/typescript.js]: You can directly use `if (this.eat(tt.dot))`

File: packages/babel-parser/src/plugins/typescript.js [hunk 3]
```diff
@@ -220,6 +220,26 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
+        node.qualifier = this.tsParseEntityName(allowReservedWords);
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/plugins/typescript.js]: Instead of getting `allowReservedWords` as a parameter, can't we just pass `true` here?

Comment (armano2) [reply packages/babel-parser/src/plugins/typescript.js]: ok will do so

File: packages/babel-parser/src/plugins/typescript.js (hunk 2 above)
Comment (armano2) [reply packages/babel-parser/src/plugins/typescript.js]: i'm kinda new to this parser, thanks for suggestion

File: packages/babel-parser/src/plugins/typescript.js [hunk 4]
```diff
@@ -220,6 +220,25 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       return result;
     }
 
+    tsParseImportType(): N.TsImportType {
+      const node: N.TsImportType = this.startNode();
+      this.expect(tt._import);
+      this.expect(tt.parenL);
+      if (!this.match(tt.string)) {
+        throw this.unexpected();
```
Comment (danez) [inline packages/babel-parser/src/plugins/typescript.js]: Can we add a nice error message here please. Something like "A TypeScript type import requires a string literal as its argument."

Comment (armano2) [reply packages/babel-parser/src/plugins/typescript.js]: typescript reports here: `String literal expected.`

Comment (armano2) [reply packages/babel-parser/src/plugins/typescript.js]: but i'm ok with adding there whatever :)

Comment (armano2) [reply packages/babel-parser/src/plugins/typescript.js]: but we should be consistent about this, and in `tsParseExternalModuleReference` there is no message.

```
      if (!this.match(tt.string)) {
        throw this.unexpected();
      }
```

Comment (armano2) [reply packages/babel-parser/src/plugins/typescript.js]: @danez what do you think about `Argument in a type import must be an String Literal`

File: packages/babel-parser/src/plugins/typescript.js [hunk 5]
```diff
@@ -225,7 +225,10 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       this.expect(tt._import);
       this.expect(tt.parenL);
       if (!this.match(tt.string)) {
-        throw this.unexpected();
+        throw this.unexpected(
+          null,
+          "Argument in a type import must be an string literal",
```
Comment (danez) [inline packages/babel-parser/src/plugins/typescript.js]: Nice, only thing it should be "... a string literal"

File: packages/babel-parser/src/plugins/typescript.js (hunk 4 above)
Comment (danez) [reply packages/babel-parser/src/plugins/typescript.js]: 👍

**Comment by Kovensky** [comment]
Thanks for following up on my PR 🙏

**Comment by hzoo** [comment]
@danez Going to make a PR tomorrow!

**Pull request merged**